packages/lit-dev-content/src/components/litdev-tutorial.ts

```diff
@@ -69,89 +86,252 @@ export class LitDevTutorial extends LitElement {
+   */
+  private get _manifest(): TutorialManifest {
+    return this._manifestTask.value ?? {steps: [{title: ''}]};
+  }
+
   /**
    * Whether the tutorial is currently in its "solved" state.
    */
   private _solved = false;
 
+  private get _projectLocation() {
+    // TODO(e111077): delete this once we go public with tutorials catalog
+    if (window.location.pathname === '/tutorial/') {
+      return 'intro-to-lit';
+    }
+    return window.location.pathname.replace('/tutorials/', '');
+  }
+
+  /**
+   * Playground project element discovered using the id from this.project
+   */
+  private get _project(): PlaygroundProject | undefined {
+    if (!this.project) {
+      return undefined;
+    }
+    return (this.getRootNode() as ShadowRoot | Document).getElementById(
+      this.project
+    ) as PlaygroundProject | undefined;
+  }
+
+  /**
+   * Generates an info from the _idx. May result in a 404 request if the index
+   * does not result in an actual file location
+   */
+  private get _info() {
+    return this._idxToInfo(this._idx);
+  }
+
+  /**
+   * Info for the next step. May result in a 404 request if the index + 1
+   * does not result in an actual file location
+   */
+  private get _nextInfo() {
+    return this._idxToInfo(this._idx + 1);
+  }
+
+  /**
+   * The root URI of the samples directory based on language preference.
+   */
+  private get _samplesRoot() {
+    return getCodeLanguagePreference() === 'ts' ? '/samples' : '/samples/js';
+  }
+
+  /**
+   * Last resolved value of the manifest task.
+   */
+  private _oldManifestValue: TutorialManifest | undefined = undefined;
+
+  /**
+   * Fetches the tutorial manifest on _projectLocation change
+   */
+  private _manifestTask = new Task(
+    this,
+    async ([projectLocation]) => {
+      const manifestRes = await fetch(
+        `${this._samplesRoot}/tutorials/${projectLocation}/tutorial.json`
+      );
+
+      return (await manifestRes.json()) as TutorialManifest;
+    },
+    () => [this._projectLocation]
+  );
+
+  /**
+   * Last resolved value of the html task.
+   */
+  private _oldHtmlValue: string | undefined = undefined;
+
+  /**
+   * Fetches the HTML-rendered markdown tutorial text on _idx change.
+   */
+  private _htmlTask = new Task(
+    this,
+    async ([idx]) => {
+      this._solved = false;
+      const active = this._info;
+
+      // Either use the preloaded html or fetch the new HTML (going back)
+      const html =
+        this._preloadedHtml?.idx === idx
+          ? await this._preloadedHtml.promise
+          : await this._fetchHtml(active.htmlSrc);
+
+      return html;
+    },
+    () => [this._idx]
+  );
+
   createRenderRoot() {
     // This is a site-specific component, and we want to inherit site-wide
     // styles.
     return this;
   }
 
+  willUpdate(): void {
+    const manifestTaskValue = this._manifestTask.value;
+
+    /**
+     * When the manifest task has just finished resolving, read the URL
+     */
+    if (
```

Comment:
This section is kind of gnarly, but I don't have an obvious suggestion to clean it up! Seems like the complexity is coming in because of having to synchronize multiple tasks that can resolve independently along with some global state like the URL. Maybe the discussion in https://github.com/lit/lit/issues/2454 will be helpful for a future refactoring!

Anyway, seems like no action needed right now. LGTM!